hoomd/hpmc/tune/nec_chain_time.py

```diff
@@ -0,0 +1,286 @@
+            particles_per_chain = chain_hit / chain_start
+        else:
+            # flake8 does not allow a single-line calculation
+            # W504 line break after binary operator
+            delta_prev = chain_hit - self.previous_hit
+            delta_start = chain_start - self.previous_start
+            particles_per_chain = delta_prev / delta_start
+
+        # We store the previous information becuase this lets us find the
+        # acceptance rate since this has last been called which allows for us to
+        # disregard the information before the last tune.
+        self.previous_hit = chain_hit
+        self.previous_start = chain_start
+        self.previous_particles_per_chain = particles_per_chain
+        return particles_per_chain
+
+    def _get_x(self):
+        return self.integrator.chain_time
+
+    def _set_x(self, value):
+        self.integrator.chain_time = value
+
+    def __hash__(self):
+        return hash(("chain_time", "", self._target, self._domain))
+
+    def __eq__(self, other):
+        return (self._target == other._target and self._domain == other._domain)
+
+
+class _InternalChainTime(_InternalAction):
+    """Internal class for the ChainTime tuner."""
+    _min_chain_time = 1e-7
+
+    def __init__(self, target, solver, max_chain_time=None):
+
+        # A flag for knowing when to update the maximum move sizes
+        self._update_chain_time = False
+
+        # A counter when tuned reaches 1 it means that the tuner has reported
+        # being tuned one time in a row. However, as the first run of the tuner
+        # is likely at timestep 0 which means that the counters are (0, 0) and
+        # _ChainTimeTuneDefinition returns y == target for that case, we need
+        # two rounds of tuning to be sure that we have converged. Since, in
+        # general, solvers do not do much if any work on already tuned tunables,
+        # this is not a performance problem.
+        self._tuned = 0
+        self._is_attached = False
+
+        self._chain_time_def = _ChainTimeTuneDefinition(
+            target, (self._min_chain_time, max_chain_time))
+
+        param_dict = ParameterDict(
+            target=OnlyTypes(float,
+                             postprocess=self._process_chain_time_target),
+            solver=SolverStep,
+            max_chain_time=OnlyTypes(
+                float,
+                allow_none=True,
+                postprocess=self._process_chain_time_range),
+            min_chain_time=OnlyTypes(
+                float, postprocess=self._process_chain_time_range))
+
+        self._param_dict.update(param_dict)
+        self.target = target
+        self.solver = solver
+        self.max_chain_time = max_chain_time
+        self.min_chain_time = self._min_chain_time
+
+    def _process_chain_time_range(self, target):
+        # check range
+        if 0 <= target:
+            pass
+        else:
+            raise ValueError(
```

Comment:
Instead of using a `pass` why not use the opposite condition and just error in that case without an else clause.